### PR TITLE
Move casemap out of experimental; final graduation cleanups

### LINKED ISSUE #3234: Stabilize the icu_casemap component
This issue tracks the work to release icu_casemap as a stable component.

Checklist (not exhaustive)

 - [x] Fix data model (https://github.com/unicode-org/icu4x/issues/3481)
 - [x] Code organization (https://github.com/unicode-org/icu4x/pull/3528)
 - [x] Expose case closure methods (https://github.com/unicode-org/icu4x/pull/3544)
 - [x] `full_fold()` functions should work with Writeable (https://github.com/unicode-org/icu4x/pull/3544)
 - [x] Add tests. Ideally upstream conformance tests (https://github.com/unicode-org/icu4x/pull/3553)
 - [x] Add docs (https://github.com/unicode-org/icu4x/pull/3553)
 - [x] Enable boilerplate lints (https://github.com/unicode-org/icu4x/pull/3544)
 - [x] Make `no_std` (https://github.com/unicode-org/icu4x/pull/3544)
 - [x] FFI
   - [x] basic FFI https://github.com/unicode-org/icu4x/pull/3663
   - [x] Add CodePointSetBuilder rather than CPILBuilder over FFI https://github.com/unicode-org/icu4x/pull/3679
   - [x] FFI tests https://github.com/unicode-org/icu4x/pull/3679
 - [x] Move data struct validation into `deserialize` to allow validation-free databake 
 - [x] Review error types, most variants seem to be datagen only
 - [x] Check if we support SpecialCasing.txt (https://github.com/unicode-org/icu4x/pull/3583)
 - [x] Make string APIs more prominent than char APIs (https://github.com/unicode-org/icu4x/pull/3584)
 - [x] Move locale to methods (except for fold/fold_turkic, https://github.com/unicode-org/icu4x/pull/3584)
 - [x] Add `titlecase_segment()` (https://github.com/unicode-org/icu4x/pull/3593)
 - [x] Handle modifiers when performing contextual operations 
 - [x] https://github.com/unicode-org/icu4x/pull/3616
 - [x] Figure out Greek accent uppercasing (https://github.com/unicode-org/icu4x/issues/3552)
 - [x] Also go through [graduation.md](https://github.com/unicode-org/icu4x/edit/main/docs/process/graduation.md) (checklist in https://github.com/unicode-org/icu4x/issues/3234#issuecomment-1636767460)
   - [x]  https://github.com/unicode-org/icu4x/pull/3689
 - [x] Maybe use more UnvalidatedStr in the data model (https://github.com/unicode-org/icu4x/pull/3690)
 - [x] Points from Markus' second review
   - [x]  Decoding should debug_assert on garbage (https://github.com/unicode-org/icu4x/pull/3759)
   - [x] split out separate `TitlecaseMapper<impl AsRef<CaseMapper>>` type, ctors: `new()`, `new_legacy()`, and then the same ones `_with_mapper` https://github.com/unicode-org/icu4x/pull/3779
   - [x] titlecase head adjustment support, with `TitleCaseOptions`, fields `HeadAdjustment { #[default] Adjust, NoAdjust }`, `TailCasing { #[default] Lowercase, PreserveCase }`  https://github.com/unicode-org/icu4x/pull/3779
   - [x] `titlecase_segment_legacy()` on CaseMapper if wanted https://github.com/unicode-org/icu4x/pull/3779
   - [x] `CaseCloser` wrapper for case closure things, similar ctors (https://github.com/unicode-org/icu4x/pull/3759)
 - [x] https://github.com/unicode-org/icu4x/issues/3801

### PR DESCRIPTION
**Pull request by Manishearth**
Part of  #3234

There is one leftover issue to be decided there, https://github.com/unicode-org/icu4x/issues/3801 , btu I'm comfortable doing the moving right now and we can fix that before release if needed.

### CODE REVIEW

File: docs/tutorials/testing/patch.toml
Comment (robertbastian) [inline docs/tutorials/testing/patch.toml:33]: nit: these are sorted by path

File: provider/datagen/src/registry.rs
Comment (robertbastian) [inline provider/datagen/src/registry.rs:142]: change this to `#[cfg(any(all(), feature = "icu_casemap"))]` for consistency. You also don't need it for both markers.

**Comment by Manishearth** [comment]
One question is: `icu_casemap` the feature got removed from datagen and the metacrate, since it's mandatory now. Should I still have a placeholder feature? or for experimental stuff do we consider the features themselves to be experimental

File: provider/datagen/src/registry.rs
Comment (Manishearth) [reply provider/datagen/src/registry.rs:142]: wait, why not on both markers?

Comment (Manishearth) [reply provider/datagen/src/registry.rs:142]: oh because of the macro

**Comment by robertbastian** [comment]
We should remove the feature, clients will need to update their code for API changes anyway.

**Review by robertbastian — APPROVED** [review]

File: components/casemap/data/Cargo.toml
Comment (robertbastian) [inline components/casemap/data/Cargo.toml:6]: I guess this should really be 1.2.0, but it isn't for any crate.

Comment (Manishearth) [reply components/casemap/data/Cargo.toml:6]: Yeah I'm trying to be consistent so that when it's publishing time we can find-and-replace easily

**Review by sffc — CHANGES_REQUESTED** [review]
https://github.com/unicode-org/icu4x/issues/3234#issuecomment-1668782445

**Comment by sffc** [comment]
> One question is: `icu_casemap` the feature got removed from datagen and the metacrate, since it's mandatory now.

I thought both datagen and the metacrate had component-specific features ?

**Comment by Manishearth** [comment]
@sffc can we please merge this without finishing the checklist? That checklist was not complete before you added more entries to it, I'm aware there's more work to be done there.

 I don't think moving this component _must_ be the last step, I'm happy to remove the "fixes ..." from the PR body but there's basically no chance 1.3 is happening without stabilizing this and we should land this somewhat churny PR at some point.

> I thought both datagen and the metacrate had component-specific features ?

**Comment by Manishearth** [comment]
I had some time and I've gone ahead and resolved _most_ of the checklist items. I'd prefer to get this merged to avoid conflicts rather than blocking this on incomplete checklists.

**Comment by sffc** [comment]
> nope, only for experimental

ok, it's not implemented yet: https://github.com/unicode-org/icu4x/issues/3165

I think we've mentioned whether we want to do this for the metacrate but I think the consensus has been that you should use the specific crates instead of the metacrate in that situation.

**Review by sffc — COMMENTED** [review]
defer to Rob on this PR. If Rob doesn't review then add me back in a few days.

**Comment by Manishearth** [comment]
autosubmit=true

**Comment by robertbastian** [comment]
There are lots of files showing as additions, I think you have `components/casemap/casemap`

**Review by robertbastian — APPROVED** [review]
Made some changes so autosubmit turned off

**Pull request merged**
